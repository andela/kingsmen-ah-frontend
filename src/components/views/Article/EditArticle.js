/* eslint-disable react/no-unused-state */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Textarea from 'react-textarea-autosize';
import PageLayout from '@components/layout/PageLayout';
import { Helmet } from 'react-helmet';
import { toast } from 'react-toastify';
import Button from '@components/commons/utilities/Button';
import Editor from './Editor';
import './index.scss';
import '../../commons/index.scss';

class EditArticle extends Component {
  static propTypes = {
    match: PropTypes.shape({
      params: PropTypes.shape({
        articleId: PropTypes.string.isRequired
      }).isRequired
    }).isRequired,
    article: PropTypes.shape({
      title: PropTypes.string,
      body: PropTypes.string,
      createdAt: PropTypes.string,
      averageRating: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number
      ]),
      author: PropTypes.shape({
        username: PropTypes.string.isRequired,
        profile: PropTypes.shape({
          firstname: PropTypes.string,
          lastname: PropTypes.string
        }).isRequired
      })
    }).isRequired,
    history: PropTypes.shape({}).isRequired,
    editArticle: PropTypes.func.isRequired,
    getSingleArticle: PropTypes.func.isRequired
  }
  constructor(props) {
    super(props);
    this.title = React.createRef();
    const { article } = this.props;

    this.state = {
      title: `${article.title}`,
    };
  }

  componentDidMount() {
    const { match: { params: {articleId} } }= this.props;
    const { getSingleArticle } = this.props;
    getSingleArticle(articleId);
  }

  onChange = (e) => {
    const { name, value } = e.target;
    this.setState({
      [name]: value
    })
  }

  getBodyObject = (body) => {
    if (!body) {
      return;
    }
    return JSON.parse(body);
  }

  editArticle = () => {
    const { match: { params: {articleId} } } = this.props;
    const { editArticle, history } = this.props
    const { title } = this.state;
    this.editor.isReady
      .then(() => {
        this.editor.save().then((outputData) => {
          const values = {
            title,
            body: JSON.stringify(outputData),
          };

          editArticle(articleId, values, history);
        });
      })
      .catch(() => {
        toast.error('Please check your internet connection!');
      });
  }

  render() {
    const { article } = this.props;
    const body = this.getBodyObject(article.body);
    
    if (!this.editor) {
      this.editor = Editor(body);
    }

    return (
      <PageLayout>
        <Helmet>
          <title>{`Editing ${article.title} - Author's Haven`}</title>
        </Helmet>
        <div className="flex float-right">
          <Button type="solid" color="blue" onClick={this.editArticle}>Edit Article</Button>
        </div>
        <div className="content-area mx-auto mt-6">
          <Textarea
            className="textarea"
            name="title"
            defaultValue={article.title}
            onChange={this.onChange}
            placeholder="Title"
            maxLength="50"
          />
          <div id="editorjs" />
        </div>
      </PageLayout>
    );
  }
}

export default EditArticle;
