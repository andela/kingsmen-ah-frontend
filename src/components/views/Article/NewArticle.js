import React, { Component } from 'react';
import PageLayout from '@components/layout/PageLayout';
import Textarea from 'react-textarea-autosize';
import { WithContext as ReactTags } from 'react-tag-input';
import { toast } from 'react-toastify';
import { Helmet } from 'react-helmet';
import { withRouter } from 'react-router-dom';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { faTimesCircle } from '@fortawesome/fontawesome-free-regular';
import Portal from '@components/commons/utilities/Portal';
import Button from '@components/commons/utilities/Button';
import FontAwesome from '../../commons/utilities/FontAwesome';
import Editor from './Editor';
import './index.scss';
import '../../commons/index.scss';

const KeyCodes = {
  comma: 188,
  enter: 13,
};
 
const delimiters = [KeyCodes.comma, KeyCodes.enter];
let ready = false;

const logger = console;
class NewArticle extends Component {
  static propTypes = {
    draft: PropTypes.bool,
    errors: PropTypes.shape({
      global: PropTypes.string
    })
  }

  static defaultProps = {
    draft: false,
    errors: ''
  }

  constructor(props) {
    super(props);
    this.title = React.createRef();
    this.state = {
      draft: false,
      title: '',
      tags: [],
      suggestions: [
        {
          id: 'Family',
          text: 'Family'
        }
      ]
    };
  }

  onChange = (e) => {
    const { name, value } = e.target;
    this.setState({
      [name]: value
    })
  }

  handleDelete = (i) => {
    const { tags } = this.state;
    this.setState({
     tags: tags.filter((tag, index) => index !== i),
    });
  }

  handleAddition = (tag) => {
    const { tags } = this.state;
    if (tags.length < 5) {
      this.setState(state => ({ tags: [...state.tags, tag] })); 
    } else {
      logger.log('The maximum number of tags allowed is 5');
    }
  }

  handleDrag = (tag, currPos, newPos) => {
    const { tags } = this.state
    const allTags = [...tags];
    const newTags = allTags.slice();

    newTags.splice(currPos, 1);
    newTags.splice(newPos, 0, tag);

    // re-render
    this.setState({ tags: newTags });
  }

  showTagModal = () => {
    const { draft } = this.props;
    this.setState({
      draft: !draft
    });
  }

  dismisModal = () => {
    this.setState({
      draft: false
    });
  }

  publishArticle = () => {
    const { createNewArticle, history } = this.props;
    const { title, tags } = this.state;

    let articleTags = [];
    tags.map(tag => {
      articleTags.push(tag.text);
    });

    this.editor.isReady
      .then(() => {
        this.editor.save().then((outputData) => {
          const values = {
            title,
            body: JSON.stringify(outputData),
            tags: articleTags
          };

          if (tags.length) {
            createNewArticle(values, history);
          } else {
            toast.error('Minimum of 1 tags required');
          }
        });
      })
      .catch(() => {
        toast.error('Please check your internet connection!');
      });
  }

  render() {
    if (!this.editor) {
      this.editor = Editor();
    }
    const { draft, tags, suggestions, title } = this.state;
    if (title.length > 4) {
      ready = true;
    }

    return (
      <PageLayout>
        <Helmet>
          <title>New Article - Author&apos;s Haven</title>
        </Helmet>
        <div className="flex float-right">
          <Button type="solid" color="blue" onClick={this.showTagModal} disabled={!ready}>Review Article</Button>
        </div>
        <div className="content-area mx-auto mt-6">
          <Textarea inputRef={this.title} className="textarea" name="title" onChange={this.onChange} placeholder="Title" data-gramm_editor="false" maxLength="50" />
          <div id="editorjs" data-gramm_editor="false" />
        </div>

        <Portal>
          <div className={classNames('relative', { 'hidden': !draft })}>
            <div className="absolute top-0 left-0 right-0 bottom bg-transparent-0 z-20 overflow-y-hidden h-screen fixed">
              <div className="flex items-center justify-center h-screen bg-gray-100">
                <div className="container mx-auto text-center text-black opacty-100 z-30">
                  <div className="float-right block">
                    <FontAwesome
                      type={faTimesCircle}
                      onClick={this.dismisModal}
                      onKeyDown={this.dismisModal}
                      styleClass="outline-none text-2xl cursor-pointer"
                    />
                  </div>
                  <div className="block">
                    <ReactTags
                      tags={tags}
                      suggestions={suggestions}
                      delimiters={delimiters}
                      handleDelete={this.handleDelete}
                      handleAddition={this.handleAddition}
                      handleDrag={this.handleDrag}
                      handleTagClick={this.handleTagClick}
                      placeholder="Add a tag"
                      inputFieldPosition="bottom"
                      classNames={{
                        tags: 'tagsClass',
                        tagInput: 'tagInputClass',
                        tagInputField: 'tag-input-field-class',
                        selected: 'selectedClass flex justify-center',
                        tag: 'tagClass',
                        remove: 'removeClass',
                        suggestions: 'suggestionsClass',
                        activeSuggestion: 'activeSuggestionClass'
                      }}
                    />
                    <Button type="solid" color="blue" onClick={this.publishArticle} id="publish-article">Publish Now</Button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </Portal>
      </PageLayout>
    );
  }
}

NewArticle.propTypes = {
  createNewArticle: PropTypes.func.isRequired,
  history: PropTypes.shape({}).isRequired
}

export default withRouter(NewArticle);
