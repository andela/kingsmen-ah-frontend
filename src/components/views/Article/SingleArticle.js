import React, { PureComponent } from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import Tags from '@components/commons/Cards/Tags';
import ReactHtmlParser from 'react-html-parser';
import Helmet from 'react-helmet';
import PageLayout from '@components/layout/PageLayout';
import ArticleRating from '@components/commons/Cards/displayStar';
import convertToJSON from '../../../helpers/convertFromJSON';
import './index.scss';

class SingleArticle extends PureComponent {
  static propTypes = {
    match: PropTypes.shape({
      params: PropTypes.shape({
        articleId: PropTypes.string.isRequired
      }).isRequired
    }).isRequired,
    tags: PropTypes.arrayOf(PropTypes.string).isRequired,
    article: PropTypes.shape({
      title: PropTypes.string,
      body: PropTypes.string,
      createdAt: PropTypes.string,
      averageRating: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number
      ]),
      author: PropTypes.shape({
        username: PropTypes.string.isRequired,
        profile: PropTypes.shape({
          firstname: PropTypes.string,
          lastname: PropTypes.string,
          avatar: PropTypes.string
        }).isRequired
      })
    }).isRequired,
    getSingleArticle: PropTypes.func.isRequired,
    getAllTags: PropTypes.func.isRequired
  }

  constructor(props) {
    super(props);

    this.defaultAvatar =
     "https://cdn2.iconfinder.com/data/icons/ios-7-icons/50/user_male2-512.png";
  }

  componentDidMount() {
    const { match: { params: {articleId} } }= this.props;
    const { getSingleArticle, getAllTags } = this.props;
    getSingleArticle(articleId);
    getAllTags(articleId);
  }

  getArticleBody = raw => {
    if (!raw) {
      return;
    }
    return convertToJSON(JSON.parse(raw));
  }

  render() {
    const { article, tags } = this.props;
    const body = this.getArticleBody(article.body);
    
    return (
      <PageLayout>
        <Helmet>
          <title>{`${article.title} - Author's Haven`}</title>
        </Helmet>
        <div className="content-area mx-auto mt-8">
          <h2 className="text-3xl font-semibold title tracking-wider">{article.title}</h2>

          <div className="my-8 flex items-center">
            <img className="w-12 h-12 rounded-full mr-4" src={article.author.profile.avatar || this.defaultAvatar} alt="Avatar of Jonathan Reinink" />
            <div className="ml-4">
              <h4 className="text-sm">Malik Berry</h4>
              <div className="flex items-center text-sm text-gray-600">
                <p className="text-xs">{moment(article.createdAt).format("MMM DD, YYYY")}</p>
                <span className="mx-3 text-black my-auto">.</span>
                <p className="text-xs">1 mins read</p>
              </div>
              <ArticleRating averageRating={article.averageRating ? article.averageRating : 0} />
            </div>
          </div>

          <div className="text-lg body">{ ReactHtmlParser(body) }</div>

          <div className="py-5 border-b-2">
            <Tags tags={tags} />
          </div>

          <div className="comments my-4">
            <h2 className="text-lg font-semibold comment-res">Responses</h2>
          </div>
        </div>
      </PageLayout>
    )
  }
}

export default SingleArticle;

