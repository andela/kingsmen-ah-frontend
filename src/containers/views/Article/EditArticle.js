import { connect } from 'react-redux';
import EditArticle from '@components/views/Article/EditArticle';
import { getSingleArticle, editArticle } from '@actions/articles';
import { withRouter } from 'react-router-dom';

const mapStateToProps = state => ({
  loading: state.article.loading,
  article: state.article.article,
  errors: state.article.errors
});

export default connect(mapStateToProps, { getSingleArticle, editArticle })(withRouter(EditArticle));
