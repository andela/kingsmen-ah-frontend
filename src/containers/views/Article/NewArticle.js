import { connect } from 'react-redux';
import NewArticle from '@components/views/Article/NewArticle';
import { createNewArticle } from '@actions/articles';

const mapStateToProps = state => ({
  isLoading: state.article.loading,
  articles: state.article.articles,
  errors: state.article.errors,
});

export default connect(mapStateToProps, { createNewArticle })(NewArticle);