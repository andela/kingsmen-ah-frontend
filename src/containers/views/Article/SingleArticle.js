import { connect } from 'react-redux';
import SingleArticle from '@components/views/Article/SingleArticle';
import { getSingleArticle, getAllTags } from '@actions/articles';

const mapStateToProps = state => ({
  isLoading: state.article.loading,
  article: state.article.article,
  tags: state.article.tags,
  errors: state.article.errors,
});

export default connect(mapStateToProps, { getSingleArticle, getAllTags })(SingleArticle);