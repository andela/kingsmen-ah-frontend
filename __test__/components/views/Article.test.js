import React from 'react';
import NewArticle from '@components/views/Article/NewArticle';
import EditArticle from '@components/views/Article/EditArticle';
import SingleArticle from '@components/views/Article/SingleArticle';

const editProps = {
    article: {
        title: 'hello',
        body: "{\"time\":1563876688871,\"blocks\":[{\"type\":\"paragraph\",\"data\":{\"text\":\"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Sed nisi lacus sed viverra tellus in hac habitasse platea. At auctor urna nunc id cursus metus aliquam<b> eleifend mi. Vulputate mi sit amet mauris commodo quis imperdiet massa tincidunt. Scelerisque varius morbi enim nunc faucibus. Turpis in eu mi bibendum neque egestas congue quisque. Sed sed risus pretium quam. Congue mauris rhoncus aenean vel elit scelerisque mauris pellentesque. Eget est lorem ipsum dolor. Hac habitasse platea dictumst vestibulum rhoncus. Duis at tellus at urna condimentum mattis pellentesque id. Facilisis gravida neque convallis a cras.</b>\"}},{\"type\":\"paragraph\",\"data\":{\"text\":\"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Sed nisi lacus sed viver<code class=\\\"inline-code\\\">ra tellus in hac ha</code>bitasse platea. At auctor urna nunc id cursus metus aliquam eleifend mi. Vulputate mi sit amet mauris commodo quis imperdiet massa tincidunt. Scelerisque varius morbi enim nunc faucibus. T<mark class=\\\"cdx-marker\\\">urpis in eu mi bibendum neque egestas congue quisque. Sed sed risus pretium quam. Congue mauris rhoncus aenean vel elit scelerisque mauris pellentesque. Eget est lorem ipsum dolor. Hac habitasse platea dictumst vestibulum rhoncus. Duis at tellus at urna condimentum mattis pellentesque id. Facilisis gravida neque convallis a cras.</mark>\"}}],\"version\":\"2.15.0\"}",
        tags: ['andela'],
        author: {
            id: 1,
            username: 'Berry',
            profile: {

            }
        }
    },
    match: {
        params: {
            articleId: 'hello'
        }
    },
    getSingleArticle: jest.fn(),
    editArticle: jest.fn(),
    getAllTags: jest.fn(),
    history: { push: jest.fn() }
};

const newProps = {
    title: 'Hello World',
    tags: ['andela']
};

describe('<Article/>', () => {
    it('renders New Article page component without crashing', () => {
        const shallowWrapper = shallow(<NewArticle {...newProps} />);

        expect(toJson(shallowWrapper)).toMatchSnapshot();
        expect(shallowWrapper.find('NewArticle')).toBeTruthy();
        expect(shallowWrapper.find('NewArticle')).toBeTruthy();
    });

    it('renders Edit Article page component without crashing', () => {
        const shallowWrapper = shallow(<EditArticle {...editProps} />);

        expect(toJson(shallowWrapper)).toMatchSnapshot();
        expect(shallowWrapper.find('EditArticle')).toBeTruthy();
    });

    it('renders Single Article page component without crashing', () => {
        const shallowWrapper = shallow(<SingleArticle {...editProps} />);

        expect(toJson(shallowWrapper)).toMatchSnapshot();
        expect(shallowWrapper.find('SingleArticle')).toBeTruthy();
    });
});