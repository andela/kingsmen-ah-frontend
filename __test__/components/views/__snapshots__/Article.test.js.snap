// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article/> renders Edit Article page component without crashing 1`] = `
<PageLayout>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Editing hello - Author's Haven
    </title>
  </HelmetWrapper>
  <div
    className="flex float-right"
  >
    <Button
      color="blue"
      onClick={[Function]}
      stretch={false}
      type="solid"
    >
      Edit Article
    </Button>
  </div>
  <div
    className="content-area mx-auto mt-6"
  >
    <TextareaAutosize
      className="textarea"
      defaultValue="hello"
      inputRef={[Function]}
      maxLength="50"
      name="title"
      onChange={[Function]}
      onHeightChange={[Function]}
      placeholder="Title"
      useCacheForDOMMeasurements={false}
    />
    <div
      id="editorjs"
    />
  </div>
</PageLayout>
`;

exports[`<Article/> renders New Article page component without crashing 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`<Article/> renders Single Article page component without crashing 1`] = `
<PageLayout>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      hello - Author's Haven
    </title>
  </HelmetWrapper>
  <div
    className="content-area mx-auto mt-8"
  >
    <h2
      className="text-3xl font-semibold title tracking-wider"
    >
      hello
    </h2>
    <div
      className="my-8 flex items-center"
    >
      <img
        alt="Avatar of Jonathan Reinink"
        className="w-12 h-12 rounded-full mr-4"
        src="https://cdn2.iconfinder.com/data/icons/ios-7-icons/50/user_male2-512.png"
      />
      <div
        className="ml-4"
      >
        <h4
          className="text-sm"
        >
          Malik Berry
        </h4>
        <div
          className="flex items-center text-sm text-gray-600"
        >
          <p
            className="text-xs"
          >
            Jul 23, 2019
          </p>
          <span
            className="mx-3 text-black my-auto"
          >
            .
          </span>
          <p
            className="text-xs"
          >
            1 mins read
          </p>
        </div>
        <DisplayStar
          averageRating={0}
        />
      </div>
    </div>
    <div
      className="text-lg body"
    >
      <div
        className="ce-block"
        key="0"
      >
        <div
          className="ce-block__content"
          key="0"
        >
          <div
            className="ce-paragraph cdx-block"
            key="0"
          >
            <p
              key="0"
            >
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Sed nisi lacus sed viverra tellus in hac habitasse platea. At auctor urna nunc id cursus metus aliquam
              <b
                key="1"
              >
                 eleifend mi. Vulputate mi sit amet mauris commodo quis imperdiet massa tincidunt. Scelerisque varius morbi enim nunc faucibus. Turpis in eu mi bibendum neque egestas congue quisque. Sed sed risus pretium quam. Congue mauris rhoncus aenean vel elit scelerisque mauris pellentesque. Eget est lorem ipsum dolor. Hac habitasse platea dictumst vestibulum rhoncus. Duis at tellus at urna condimentum mattis pellentesque id. Facilisis gravida neque convallis a cras.
              </b>
            </p>
          </div>
        </div>
      </div>
      <div
        className="ce-block"
        key="1"
      >
        <div
          className="ce-block__content"
          key="0"
        >
          <div
            className="ce-paragraph cdx-block"
            key="0"
          >
            <p
              key="0"
            >
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Sed nisi lacus sed viver
              <code
                className="inline-code"
                key="1"
              >
                ra tellus in hac ha
              </code>
              bitasse platea. At auctor urna nunc id cursus metus aliquam eleifend mi. Vulputate mi sit amet mauris commodo quis imperdiet massa tincidunt. Scelerisque varius morbi enim nunc faucibus. T
              <mark
                className="cdx-marker"
                key="3"
              >
                urpis in eu mi bibendum neque egestas congue quisque. Sed sed risus pretium quam. Congue mauris rhoncus aenean vel elit scelerisque mauris pellentesque. Eget est lorem ipsum dolor. Hac habitasse platea dictumst vestibulum rhoncus. Duis at tellus at urna condimentum mattis pellentesque id. Facilisis gravida neque convallis a cras.
              </mark>
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="py-5 border-b-2"
    >
      <Tags />
    </div>
    <div
      className="comments my-4"
    >
      <h2
        className="text-lg font-semibold comment-res"
      >
        Responses
      </h2>
    </div>
  </div>
</PageLayout>
`;
